File: codeforces/1800~1899/1858/b.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0, m = 0, d = 0;
  cin >> n >> m >> d;

  vector<int> s(m);
  for (int i = 0; i < m; ++i) cin >> s[i];

  s.insert(s.begin(), 1);

  vector<int> p = {1};
  for (int i = 1; i < s.size(); ++i) {
    int diff = s[i] - s[i-1];
    int eat = (diff + d - 1) / d;
    p.push_back(p.back() + eat);
  }

  if (s.back() != n) {
    p.push_back(p.back() + (n - s.back()) / d);
  } else {
    p.push_back(p.back());
  }

  int size = s.size();
  // cout << "p=";
  // for (int i = 0; i < p.size(); ++i) cout << p[i] << " ";
  // cout << endl;

  int min_val = numeric_limits<int>::max();
  int amount = 0;
  // try to remove i seller
  for (int i = 1; i < size; ++i) {
    int result = 0;
    if (i == size-1) {
      int diff = n - s[i-1];
      int eat = diff / d;
      result = p[i-1] + eat;
    } else {
      int end = s[i+1];
      int diff = end - s[i-1];
      int eat = (diff + d - 1) / d;
      result = p[i-1] + p.back() - p[i+1] + eat;
    }

    // cout << "remove=" << i << " , result=" << result << endl;
    if (result < min_val) {
      min_val = result;
      amount = 1;
    } else if (result == min_val) {
      amount++;
    }
  }

  cout << min_val << " " << amount << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1800~1899/1858/b.rs
use std::io;
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let line = read_vec::<i32>();
    let n = line[0];
    let d = line[2];

    let mut s = read_vec::<i32>();
    s.insert(0, 1);

    let mut p: Vec<i32> = vec![1];
    for i in 1..s.len() {
        let diff = s[i] - s[i-1];
        let eat = (diff + d - 1) / d;
        p.push(*p.last().unwrap() + eat);
    }

    if s[s.len()-1] != n {
        p.push(*p.last().unwrap() + (n - *s.last().unwrap()) / d);
    } else {
        p.push(*p.last().unwrap());
    }

    let mut min_val: i32 = std::i32::MAX;
    let mut amount: i32 = 0;

    for i in 1..s.len() {
        let result: i32;
        if i == s.len()-1 {
            let diff = n - s[i-1];
            let eat = diff / d;
            result = p[i-1] + eat;
        } else {
            let diff = s[i+1] - s[i-1];
            let eat = (diff + d - 1) / d;
            result = p[i-1] + *p.last().unwrap() - p[i+1] + eat;
        }

        if result < min_val {
            min_val = result;
            amount = 1;
        } else if result == min_val {
            amount += 1;
        }
    }

    println!("{} {}", min_val, amount);
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
